File: py/todoosy/scheme.py
"""
Todoosy Scheme Parser
"""

import re
from .types import Scheme

TIMEZONE_HEADING_REGEX = re.compile(r'^#\s+Timezone\s*$', re.IGNORECASE)
PRIORITIES_HEADING_REGEX = re.compile(r'^#\s+Priorities\s*$', re.IGNORECASE)
MISC_HEADING_REGEX = re.compile(r'^#\s+Misc\s*$', re.IGNORECASE)
CALENDAR_FORMAT_HEADING_REGEX = re.compile(r'^#\s+Calendar\s+Format\s*$', re.IGNORECASE)
FORMATTING_STYLE_HEADING_REGEX = re.compile(r'^#\s+Formatting\s+Style\s*$', re.IGNORECASE)
PRIORITY_LINE_REGEX = re.compile(r'^[*\-]?\s*P(\d+)\s*[-–—]\s*(.+)$', re.IGNORECASE)

VALID_CALENDAR_FORMATS = {'yyyy-mm-dd', 'yyyy/mm/dd', 'mm/dd/yyyy', 'dd/mm/yyyy'}
VALID_FORMATTING_STYLES = {'roomy', 'balanced', 'tight'}


def parse_scheme(text: str) -> Scheme:
    """Parse a todoosy scheme file."""
    lines = text.split('\n')
    scheme = Scheme(timezone=None, priorities={}, misc='todoosy.md/Misc', calendar_format='yyyy-mm-dd', formatting_style='roomy')

    current_section = 'none'
    calendar_format_set = False
    formatting_style_set = False

    for line in lines:
        trimmed = line.strip()

        # Check for section headers
        if TIMEZONE_HEADING_REGEX.match(trimmed):
            current_section = 'timezone'
            continue

        if PRIORITIES_HEADING_REGEX.match(trimmed):
            current_section = 'priorities'
            continue

        if MISC_HEADING_REGEX.match(trimmed):
            current_section = 'misc'
            continue

        if CALENDAR_FORMAT_HEADING_REGEX.match(trimmed):
            current_section = 'calendar_format'
            continue

        if FORMATTING_STYLE_HEADING_REGEX.match(trimmed):
            current_section = 'formatting_style'
            continue

        # Check if we hit another heading (any level)
        if re.match(r'^#+\s+', trimmed) and current_section != 'none':
            current_section = 'none'
            continue

        # Skip empty lines
        if not trimmed:
            continue

        # Process content based on current section
        if current_section == 'timezone':
            if scheme.timezone is None:
                scheme.timezone = trimmed

        elif current_section == 'priorities':
            match = PRIORITY_LINE_REGEX.match(trimmed)
            if match:
                level = match.group(1)
                label = match.group(2).strip()
                scheme.priorities[level] = label
            else:
                # Try format without bullet
                alt_match = re.match(r'^P(\d+)\s*[-–—]\s*(.+)$', trimmed, re.IGNORECASE)
                if alt_match:
                    level = alt_match.group(1)
                    label = alt_match.group(2).strip()
                    scheme.priorities[level] = label

        elif current_section == 'misc':
            # First non-empty line after Misc heading is the misc location (filename/headingname)
            if trimmed and not trimmed.startswith('#'):
                scheme.misc = trimmed

        elif current_section == 'calendar_format':
            # First non-empty line after Calendar Format heading is the format
            if not calendar_format_set and trimmed and not trimmed.startswith('#'):
                scheme.calendar_format = trimmed.lower()
                calendar_format_set = True

        elif current_section == 'formatting_style':
            # First non-empty line after Formatting Style heading is the style
            if not formatting_style_set and trimmed and not trimmed.startswith('#'):
                scheme.formatting_style = trimmed.lower()
                formatting_style_set = True

    return scheme

File: js/src/scheme.ts
/**
 * Todoosy Scheme Parser
 */

import type { Scheme } from './types.js';

const TIMEZONE_HEADING_REGEX = /^#\s+Timezone\s*$/i;
const PRIORITIES_HEADING_REGEX = /^#\s+Priorities\s*$/i;
const MISC_HEADING_REGEX = /^#\s+Misc\s*$/i;
const CALENDAR_FORMAT_HEADING_REGEX = /^#\s+Calendar\s+Format\s*$/i;
const FORMATTING_STYLE_HEADING_REGEX = /^#\s+Formatting\s+Style\s*$/i;
const PRIORITY_LINE_REGEX = /^[*\-]?\s*P(\d+)\s*[-–—]\s*(.+)$/i;

export const VALID_CALENDAR_FORMATS = new Set(['yyyy-mm-dd', 'yyyy/mm/dd', 'mm/dd/yyyy', 'dd/mm/yyyy']);
export const VALID_FORMATTING_STYLES = new Set(['roomy', 'balanced', 'tight']);

export function parseScheme(text: string): Scheme {
  const lines = text.split('\n');
  const scheme: Scheme = {
    timezone: null,
    priorities: {},
    misc: 'todoosy.md/Misc',
    calendar_format: 'yyyy-mm-dd',
    formatting_style: 'roomy',
  };

  let currentSection: 'none' | 'timezone' | 'priorities' | 'misc' | 'calendar_format' | 'formatting_style' = 'none';
  let calendarFormatSet = false;
  let formattingStyleSet = false;

  for (const line of lines) {
    const trimmed = line.trim();

    // Check for section headers
    if (TIMEZONE_HEADING_REGEX.test(trimmed)) {
      currentSection = 'timezone';
      continue;
    }

    if (PRIORITIES_HEADING_REGEX.test(trimmed)) {
      currentSection = 'priorities';
      continue;
    }

    if (MISC_HEADING_REGEX.test(trimmed)) {
      currentSection = 'misc';
      continue;
    }

    if (CALENDAR_FORMAT_HEADING_REGEX.test(trimmed)) {
      currentSection = 'calendar_format';
      continue;
    }

    if (FORMATTING_STYLE_HEADING_REGEX.test(trimmed)) {
      currentSection = 'formatting_style';
      continue;
    }

    // Check if we hit another heading (any level)
    if (/^#+\s+/.test(trimmed) && currentSection !== 'none') {
      currentSection = 'none';
      continue;
    }

    // Skip empty lines
    if (trimmed === '') {
      continue;
    }

    // Process content based on current section
    if (currentSection === 'timezone') {
      // First non-empty line after Timezone heading is the timezone
      if (scheme.timezone === null) {
        scheme.timezone = trimmed;
      }
    } else if (currentSection === 'priorities') {
      // Try to parse priority line
      const match = trimmed.match(PRIORITY_LINE_REGEX);
      if (match) {
        const level = match[1];
        const label = match[2].trim();
        scheme.priorities[level] = label;
      } else if (!trimmed.startsWith('#')) {
        // Also try format without bullet: "P0 - Label"
        const altMatch = trimmed.match(/^P(\d+)\s*[-–—]\s*(.+)$/i);
        if (altMatch) {
          const level = altMatch[1];
          const label = altMatch[2].trim();
          scheme.priorities[level] = label;
        }
      }
    } else if (currentSection === 'misc') {
      // First non-empty line after Misc heading is the misc location (filename/headingname)
      if (trimmed && !trimmed.startsWith('#')) {
        scheme.misc = trimmed;
      }
    } else if (currentSection === 'calendar_format') {
      // First non-empty line after Calendar Format heading is the format
      if (!calendarFormatSet && trimmed && !trimmed.startsWith('#')) {
        scheme.calendar_format = trimmed.toLowerCase();
        calendarFormatSet = true;
      }
    } else if (currentSection === 'formatting_style') {
      // First non-empty line after Formatting Style heading is the style
      if (!formattingStyleSet && trimmed && !trimmed.startsWith('#')) {
        scheme.formatting_style = trimmed.toLowerCase();
        formattingStyleSet = true;
      }
    }
  }

  return scheme;
}
